song/song_.py

```diff
@@ -161,13 +161,15 @@ def fit(self, X: np.array) -> None:
         is_execute = False
         ephoc = 0
         while not is_execute:
-            for x in np.random.permutation(X):
+            for idx in np.random.permutation(self.n_in_data_num):
+                self.idx = idx
+                x = self.X[idx]
                 self._update_neighbors(x)
                 i_1 = self.neighbor_idxs[0]
                 old_connecteds = set(self.topology[i_1].keys())
                 self._edge_curation()
-                connecteds = self.topology[i_1].keys()
-                if old_connecteds == connecteds:
+                connecteds = set(self.topology[i_1].keys())
+                if old_connecteds == connecteds and ephoc > 0:
```

Comment:
Under the first epoch (when epoch == 0), always this condition, so exclude the case.